### PR TITLE
Fix security issues of the argument parser

### PR DESCRIPTION
**Pull request by smehringer**
Addresses several things of #1101 

Best reviewd commit by commit.

### CODE REVIEW

File: include/seqan3/argument_parser/detail/version_check.hpp [hunk 1]
```diff
@@ -88,27 +90,12 @@ class version_checker
 #endif
         std::smatch versionMatch;
 
+        // Ensure version string is not corrupt
         if (!version_.empty() && /*regex allows version prefix instead of exact match */
             std::regex_search(version_, versionMatch, std::regex("^([[:digit:]]+\\.[[:digit:]]+\\.[[:digit:]]+).*")))
         {
             version = versionMatch.str(1); // in case the git revision number is given take only version number
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/argument_parser/detail/version_check.hpp]: (note the underscore difference) 
if `version_ == "some malicious code not matching the regex"` then the member `version` will never be set inside the if clause and remains the default `"0.0.0"`.

Is the "zusammenploebbeln" of the command more clear now with the changes?

[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/argument_parser/detail/version_check.hpp]: AH okay, I didn't catch the underscore.

> Is the "zusammenploebbeln" of the command more clear now with the changes?

:+1:

**Comment by marehr** [comment]
Hmm jenkins and travis failing.

> home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp: In member function ‘virtual void parser_design_error_app_name_validation_Test::TestBody()’:
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:18:138: error: expected primary-expression before ‘parser’
     EXPECT_NO_THROW((argument_parser parser{"test_parser", 1, argv}));
                                                                                                                                          ^     
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:18:138: error: expected ‘)’ before ‘parser’
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:19:138: error: expected primary-expression before ‘parser’
     EXPECT_NO_THROW((argument_parser parser{"test-parser1234_foo", 1, argv}));
                                                                                                                                          ^     
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:19:138: error: expected ‘)’ before ‘parser’
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:21:189: error: expected primary-expression before ‘parser’
     EXPECT_THROW((argument_parser parser{"test parser", 1, argv}),       parser_design_error);
                                                                                                                                                                                             ^     
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:21:189: error: expected ‘)’ before ‘parser’
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:22:189: error: expected primary-expression before ‘parser’
     EXPECT_THROW((argument_parser parser{"test;", 1, argv}),             parser_design_error);
                                                                                                                                                                                             ^     
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:22:189: error: expected ‘)’ before ‘parser’
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:23:189: error: expected primary-expression before ‘parser’
     EXPECT_THROW((argument_parser parser{";", 1, argv}),                 parser_design_error);
                                                                                                                                                                                             ^     
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:23:189: error: expected ‘)’ before ‘parser’
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:24:189: error: expected primary-expression before ‘parser’
     EXPECT_THROW((argument_parser parser{"test;bad script:D", 1, argv}), parser_design_error);
                                                                                                                                                                                             ^     
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:24:189: error: expected ‘)’ before ‘parser’
/home/travis/build/seqan/seqan3/test/unit/argument_parser/argument_parser_design_error_test.cpp:16:18: error: unused variable ‘argv’ [-Werror=unused-variable]
     const char * argv[] = {"./argument_parser_test"};
                  ^~~~

File: test/unit/argument_parser/argument_parser_design_error_test.cpp [hunk 2]
```diff
@@ -11,6 +11,19 @@
 
 using namespace seqan3;
 
+TEST(parser_design_error, app_name_validation)
+{
+    const char * argv[] = {"./argument_parser_test"};
```
Comment (marehr) [inline test/unit/argument_parser/argument_parser_design_error_test.cpp]: I think the error is caused by the `{}`

File: test/unit/argument_parser/argument_parser_design_error_test.cpp [hunk 3]
```diff
@@ -11,6 +11,19 @@
...
+
+    EXPECT_NO_THROW((argument_parser parser{"test_parser", 1, argv}));
```
Comment (marehr) [inline test/unit/argument_parser/argument_parser_design_error_test.cpp]: I think you should do this:

```suggestion
    EXPECT_NO_THROW((argument_parser{"test_parser", 1, argv}));
```

Comment (marehr) [reply test/unit/argument_parser/argument_parser_design_error_test.cpp]: Or
```suggestion
    EXPECT_NO_THROW(([]()
    {
        argument_parser parser{"test_parser", 1, argv};
    }));
```

File: include/seqan3/argument_parser/argument_parser.hpp [hunk 4]
```diff
@@ -166,6 +167,8 @@ class argument_parser
      * \param[in] argv           The command line arguments to parse.
      * \param[in] version_check  Notify users about app version updates (default true).
      *
+     * \throws seqan3::parser_design_error if the application name contains illegal characters.
```
Comment (rrahn) [inline include/seqan3/argument_parser/argument_parser.hpp]: is somewhere documented what the legal/illegal characters are?

File: include/seqan3/argument_parser/argument_parser.hpp [hunk 5]
```diff
@@ -166,6 +167,10 @@ class argument_parser
...
+     *
+     * The application name must only contain alpha-numeric characters or '_' and '-' (regex: \"^[a-zA-Z0-9_-]+$\").
```
Comment (rrahn) [inline include/seqan3/argument_parser/argument_parser.hpp]:
```suggestion
     * The application name must only contain alpha-numeric characters, '_' or '-', i.e. the following regex must evaluate to true: `\"^[a-zA-Z0-9_-]+$\"`.
```

**Comment by smehringer** [comment]
@rrahn Please check the first commit and the last again, I had to rebase after your timeout change was merged.

@h-2 Does `/usr/bin/env -i` work on FreeBSD/OpenBSD ?

File: include/seqan3/argument_parser/detail/version_check.hpp [hunk 6]
```diff
@@ -429,10 +447,10 @@ class version_checker
...
         return "powershell.exe -NoLogo -NonInteractive -Command \"& {Invoke-WebRequest -erroraction 'silentlycontinue' -OutFile";
 #else  // Unix based platforms.
-        if (!system("wget --version > /dev/null 2>&1"))
-            return "wget --timeout=10 --tries=1 -q -O";
-        else if (!system("curl --version > /dev/null 2>&1"))
-            return "curl --connect-timeout 10 -o";
+        if (!system("/usr/bin/env -i wget --version > /dev/null 2>&1"))
+            return "/usr/bin/env -i wget --timeout=10 --tries=1 -q -O";
+        else if (!system("/usr/bin/env -i curl --version > /dev/null 2>&1"))
+            return "/usr/bin/env -i curl --connect-timeout 10 -o";
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/version_check.hpp]: die commands für openbsd und freebsd müssten auch noch mit env erweitert werden. Beide Systeme haben auch das env tool.

**Pull request merged**